liquid-compiler/src/parser.rs

```diff
@@ -3,477 +3,831 @@
+    fn unwrap_variable(&mut self) -> std::result::Result<Pair<'a>, ()> {
+        let value = self.unwrap_value()?;
+
+        let variable = value
+            .into_inner()
+            .next()
+            .expect("A value is made of one token.");
 
-        let result = parse_expression(&tokens, &null_options()).unwrap();
-        let result = result.render(&mut context).unwrap();
-        assert_eq!("42", result);
+        if variable.as_rule() != Rule::Variable {
+            return Err(());
+        }
+
+        Ok(variable)
     }
 
-    #[test]
-    fn array_index_access() {
-        let tokens = granularize("post[0]").unwrap();
-        let mut context = Context::new();
-        let mut post = Array::new();
-        post.push(Value::scalar(42i32));
-        context.stack_mut().set_global("post", Value::Array(post));
+    fn unwrap_identifier(&mut self) -> std::result::Result<Pair<'a>, ()> {
+        let variable = self.unwrap_variable()?;
 
-        let result = parse_expression(&tokens, &null_options()).unwrap();
-        let result = result.render(&mut context).unwrap();
-        assert_eq!("42", result);
+        let mut indexes = variable.into_inner();
+        let identifier = indexes
+            .next()
+            .expect("Unwrapping identifier out of variable.");
+        if indexes.next().is_some() {
+            // There are indexes: it can't be a value
+            return Err(());
+        }
+
+        Ok(identifier)
     }
 
-    #[test]
-    fn mixed_access() {
-        let tokens = granularize("post.child[0]").unwrap();
-        let mut context = Context::new();
-        let mut post = Object::new();
-        post.insert("child".into(), Value::Array(vec![Value::scalar(42i32)]));
-        context.stack_mut().set_global("post", Value::Object(post));
+    fn unwrap_literal(&mut self) -> std::result::Result<Pair<'a>, ()> {
+        let value = self.unwrap_value()?;
+
+        let literal = value
+            .into_inner()
+            .next()
+            .expect("A value is made of one token.");
+
+        if literal.as_rule() != Rule::Literal {
+            return Err(());
+        }
+
+        Ok(literal)
+    }
+
+    /// Tries to obtain a `FilterChain` from this token.
+    pub fn expect_filter_chain(mut self) -> std::result::Result<FilterChain, Self> {
+        let filterchain = self.unwrap_filter_chain().map_err(|_| {
+            self.expected.push(Rule::FilterChain);
+            self
+        })?;
 
-        let result = parse_expression(&tokens, &null_options()).unwrap();
-        let result = result.render(&mut context).unwrap();
-        assert_eq!("42", result);
+        Ok(parse_filter_chain(filterchain))
+    }
+
+    /// Tries to obtain a value from this token.
+    ///
+    /// Do not confuse this value with `liquid-value`'s `Value`.
+    /// In this context, value refers to either a literal value or a variable.
+    pub fn expect_value(mut self) -> std::result::Result<Expression, Self> {
```

Comment:
So, the problem of expecting something is that sometimes even if one `expect_` function fails you might want to try another. For example, `cycle_tag.rs`, line 58:

```

Reply:
Ok, so this is trying to imitate nom's `IResult`.  It then modifies `self` with the extra information that may be desired in an error.

My concern is that it feels a little dirty.  I try to stick to a rule of the `Result::Err` always having `impl Error`.

So looking into this more:

- Does `expect` taking a `self` buy us anything?  I'm assuming there isn't any allocated memory we're able to reuse.  If it instead took `&self`, then we could make multiple calls without chaining through the error
- Part of the challenge here is getting pest's helpful errors without pest being in the public API.  If we skipped their nice errors, then we wouldn't have a problem.  If instead we exposed `Rule`, then we wouldn't have a problem.  Seems like there is the potential for a middle ground route where we create our own error using the information pest provides us.
- We could switch from `Result` to `Either` or a custom enum.
- We could hold off on any significant change for now to see how the parser evolves.

What are your thoughts?

Reply:
How about we switch to `Either` or a custom enum for now to better convey the intent and worry about changing up error handling later?

Reply:
I made a custom enum `TryMatchToken`, because I'm not a big fan of `Either`. It seems very ambiguous regarding the meaning of the type.

I think this will take care of this issue (for now).